[receiver/kafkareceiver] [internal/kafka] Add support for configuring Kafka consumer rebalance strategy and group instance ID

bvsvas opened the pull request:
As metioned in issue# [39513](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/39513) 

… Kafka consumer rebalance strategy and group instance ID

This enhancement introduces two optional settings: group_rebalance_strategy and group_instance_id.
  These allow users to override the default Range-based rebalance strategy and optionally provide a static instance ID (as per KIP-345) for cooperative sticky balancing.
  This is particularly useful when handling high-cardinality metric workloads, as it reduces rebalance impact, improves cache reuse, and boosts CPU efficiency.
  Both settings are optional to maintain full backward compatibility.

<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes

<!--Describe what testing was performed and which tests were added.-->
#### Testing

<!--Describe the documentation added.-->
#### Documentation

<!--Please delete paragraphs that you did not use before submitting.-->

srinivasvenkatabevara committed:
[receiver/kafkareceiver] [internal/kafka] Add support for configuring Kafka consumer rebalance strategy and group instance ID

This enhancement introduces two optional settings: group_rebalance_strategy and group_instance_id.
  These allow users to override the default Range-based rebalance strategy and optionally provide a static instance ID (as per KIP-345) for cooperative sticky balancing.
  This is particularly useful when handling high-cardinality metric workloads, as it reduces rebalance impact, improves cache reuse, and boosts CPU efficiency.
  Both settings are optional to maintain full backward compatibility.

atoulme committed:
Merge branch 'main' into kafka-sticky-balance

atoulme commented on internal/kafka/configkafka/config.go:
```diff
@@ -98,6 +98,14 @@ type ConsumerConfig struct {
+	// RebalanceStrategy specifies the strategy to use for partition assignment.
+	// Possible values are "range", "roundrobin", and "sticky".
+	// Defaults to "range".
+	GroupRebalanceStrategy string `mapstructure:"group_rebalance_strategy" validate:"omitempty,oneof=range roundrobin sticky"`
```
we don't use the `validate` annotation.

atoulme commented on internal/kafka/client.go:
```diff
@@ -125,3 +133,20 @@ func newSaramaClientConfig(ctx context.Context, config configkafka.ClientConfig)
 }
+
+func rebalanceStrategy(strategy string) sarama.BalanceStrategy {
+	if len(strategy) == 0 {
+		return nil
+	}
```
isn't the same as line 150?

axw commented on internal/kafka/client_test.go:
```diff
@@ -313,3 +313,75 @@ func TestNewSaramaClient_TLS(t *testing.T) {
 	})
 }
+
+func TestNewSaramaConsumerGroup_RebalanceAndInstanceId(t *testing.T) {
```
The two configuration settings are independent, so should we test them independently too?

axw replied:
Right, but is it important that they are tested together? It seems to me we could make this one just about testing GroupRebalanceStrategy

axw reviewed (commented):
Looks great, thanks @bvsvas! Just a minor comment, on top of what Antoine has already mentioned.

srinivasvenkatabevara committed:
[receiver/kafkareceiver] [internal/kafka] Add support for configuring Kafka consumer rebalance strategy and group instance ID

1) Extra Unit tests for Client
2) Removed validate from config

srinivasvenkatabevara commented:
> Looks great, thanks @bvsvas! Just a minor comment, on top of what Antoine has already mentioned.

Thanks @axw for the review, I have done requested changes.

axw commented on internal/kafka/client.go:
```diff
@@ -125,3 +133,20 @@ func newSaramaClientConfig(ctx context.Context, config configkafka.ClientConfig)
 }
+
+func rebalanceStrategy(strategy string) sarama.BalanceStrategy {
+	if len(strategy) == 0 {
+		return nil
+	}
```
```suggestion
```

It's already covered by the default case below

axw replied:
@bvsvas this one is still pending.

axw commented on internal/kafka/client_test.go:
```diff
@@ -313,3 +313,178 @@ func TestNewSaramaClient_TLS(t *testing.T) {
+			protocolVersion: "2.2.0",
+			expectedErr:     "Consumer.Group.InstanceId need Version >= 2.3",
+		},
+		{
+			name:            "No GroupInstanceID with Invalid Protocol Version",
+			groupInstanceID: "",
+			protocolVersion: "2.2.0",
+			expectedErr:     "",
+		},
```
```suggestion
```

Since GroupInstanceID won't be set, this doesn't seem like a relevant test case. Can we remove this from here? If we do that, perhaps this doesn't need to be a table test.

srinivasvenkatabevara replied:
No error test scenario, in this case. 

With same version "2.2.0", having groupInstanceID will result into error and without groupInstanceID will not result into error.

axw replied:
Understood, just seems a bit unnecessary given that we would also test this behaviour in all other consumer group tests that don't involve setting GroupInstanceID.

axw reviewed (approved):
Thanks @bvsvas. LGTM. I would like to see the tests simplified slightly more, but it's not a big deal.

pavolloffay reviewed (commented):
Could you please document these options in the readme as well?

srinivasvenkatabevara commented:
> Could you please document these options in the readme as well?

Thanks @pavolloffay for the review, that's good point, I will update.

srinivasvenkatabevara committed:
Add support for configuring Kafka consumer rebalance strategy and group instance ID

Updated readme with new configuration parameters group_rebalance_strategy and group_instance_id

srinivasvenkatabevara commented:
> > Could you please document these options in the readme as well?
> 
> Thanks @pavolloffay for the review, that's good point, I will update.

Done.

srinivasvenkatabevara commented:
@pavolloffay - Could you please let me know if there’s anything pending? Are we good to merge this change?

srinivasvenkatabevara committed:
Add support for configuring Kafka consumer rebalance strategy and group instance ID

Removed extra check

srinivasvenkatabevara commented:
Are we good to merge this change?

atoulme commented:
Please address the conflict.

srinivasvenkatabevara committed:
Merge remote-tracking branch 'origin/main' into kafka-sticky-balance

srinivasvenkatabevara commented:
> Please address the conflict.

Completed

atoulme commented:
Please fix the CI

srinivasvenkatabevara committed:
[internal/kafka] client test format fix

bvsvas committed:
Merge branch 'main' into kafka-sticky-balance

srinivasvenkatabevara commented:
> Please fix the CI

Done

vashisthaa committed:
[receiver/KafkaReceiver] Update README.md

vs667919 committed:
[receiver/kafkareceiver] Update README.md

srinivasvenkatabevara committed:
Merge branch 'main' into kafka-sticky-balance

srinivasvenkatabevara commented:
Hope all good here? Are we good to merge this change?

srinivasvenkatabevara committed:
Merge branch 'main' into kafka-sticky-balance

bvsvas committed:
Merge branch 'main' into kafka-sticky-balance

srinivasvenkatabevara commented:
@atoulme @axw - Just wanted to follow up on this PR. Let me know if there’s anything I should update.
Thanks for your time!
